Port to Minecraft 1.21.1

sennecools opened the pull request:
## Summary

This PR ports ReFramed from Minecraft 1.20.4 to 1.21.1, updating all dependencies and migrating to new APIs introduced in the latest Minecraft/Fabric versions.

## Changes

### Build System
- Updated Minecraft from 1.20.4 to 1.21.1
- Updated Fabric API to 0.116.7+1.21.1
- Updated Fabric Loader to 0.16.9
- Upgraded Java toolchain from 17 to 21
- Updated Gradle to 8.12 and Loom to 1.9.2

### API Migrations

**Model Loading API** - Complete rewrite for new system:
- Migrated from deprecated ModelResourceProvider/ModelVariantProvider to ModelLoadingPlugin
- Implemented resolveModel() and modifyModelOnLoad() callbacks
- Updated handling of ModelIdentifier (now a record type)

**Component Data System** - Replaced NBT-based ItemStack data:
- Migrated from getBlockEntityNbt() to DataComponentTypes.BLOCK_ENTITY_DATA
- Updated all item stack NBT access to use Component API
- Converted blueprint and item tooltip systems

**BlockEntity Serialization**:
- Updated readNbt()/writeNbt() signatures to include RegistryWrapper.WrapperLookup
- Changed method visibility from public to protected

**Block Interaction** - Critical signature change:
- Updated onUse() method across all interactive blocks
- Removed Hand parameter (removed in 1.21.1)
- Changed visibility from public to protected
- Implemented hand-loop workaround in helper methods

**Other API Updates**:
- Replaced new Identifier() with Identifier.of() and Identifier.ofVanilla()
- Updated UnbakedModel.bake() signature
- Updated data generators to include CompletableFuture<WrapperLookup> parameter
- Updated WeightedCollection.getData() to .data()

### Compatibility Modules

Temporarily disabled the following integrations until compatible 1.21.1 versions are available:
- Continuity (connected textures)
- Sodium (render optimization)
- Indium (Fabric rendering API)
- Athena (connected textures)
- Axiom (world editor)

These modules have been moved to compat_disabled and will be re-enabled in future updates.

## Testing

All core features have been tested and verified working:
- ✅ Block placement and rendering
- ✅ Camo application (right-click with blocks)
- ✅ Light emission toggle (glowstone dust)
- ✅ Redstone emission toggle (redstone torch)
- ✅ Collision toggle (popped chorus fruit)
- ✅ Screwdriver rotation (directional blocks)
- ✅ Blueprint copying
- ✅ Hammer functionality
- ✅ All frame block types and variants

## Build Requirements

This version requires **Java 21** to build and run:

\`\`\`bash
export JAVA_HOME=/usr/lib/jvm/java-21-openjdk
./gradlew build
\`\`\`

## Future Work

- Re-enable Continuity integration for connected textures support once compatible version is available
- Re-enable other compatibility modules as their 1.21.1 versions are released

## Additional Documentation

See PORTING_NOTES.md for comprehensive technical details on all changes made during the port.

sennecools committed:
build: update to Minecraft 1.21.1 and Java 21

- Update Minecraft version from 1.20.4 to 1.21.1
- Update Yarn mappings to 1.21.1+build.3
- Update Fabric Loader to 0.16.9
- Update Fabric API to 0.116.7+1.21.1
- Upgrade Java toolchain from 17 to 21
- Update Gradle wrapper to 8.12
- Update Fabric Loom to 1.9.2
- Add Java 21 toolchain to buildSrc to prevent version conflicts

sennecools committed:
refactor: migrate to new Model Loading Plugin API

Replace deprecated ModelResourceProvider and ModelVariantProvider with
the new ModelLoadingPlugin system introduced in 1.21.1.

- Use resolveModel callback for block model resolution
- Use modifyModelOnLoad callback for item model interception
- Handle ModelIdentifier as record type instead of Identifier subclass
- Update UnbakedModel.bake() signature (removed Identifier parameter)
- Update all custom unbaked model implementations

sennecools committed:
refactor: update BlockEntity NBT serialization for 1.21.1

Update readNbt and writeNbt methods to match new 1.21.1 signatures:
- Add RegistryWrapper.WrapperLookup parameter to all NBT methods
- Change visibility from public to protected
- Update toInitialChunkDataNbt() to include WrapperLookup parameter
- Migrate ItemStack NBT access to Component data system

sennecools committed:
refactor: migrate from NBT to Component data system

Replace deprecated NBT-based ItemStack data access with the new
Component data system introduced in 1.21.1.

- Replace getBlockEntityNbt() with DataComponentTypes.BLOCK_ENTITY_DATA
- Use NbtComponent.copyNbt() for NBT access from components
- Update BlockItemMixin injection to use component API
- Migrate blueprint item NBT handling to components
- Update tooltip methods to new signature

sennecools committed:
fix: update block interaction methods for 1.21.1 API changes

Update onUse method signature across all interactive blocks:
- Remove Hand parameter (removed in 1.21.1)
- Change visibility from public to protected
- Implement hand-loop workaround in helper methods to try both hands
- Update all block classes: doors, trapdoors, fences, buttons, etc.

This fixes camo application and all block interactions that were
broken by the API change.

sennecools committed:
refactor: update data generators for 1.21.1

Update all FabricDataProvider implementations to include
CompletableFuture<RegistryWrapper.WrapperLookup> parameter:
- Language provider
- Block loot tables
- Recipes
- Advancements

sennecools committed:
refactor: update miscellaneous API calls for 1.21.1

- Replace new Identifier() with Identifier.of() and Identifier.ofVanilla()
- Update WeightedCollection.getData() to .data()
- Clean up import statements
- Update screwdriver item implementation

sennecools committed:
refactor: temporarily disable compatibility modules

Move compatibility mixins to compat_disabled directory until
compatible versions of integration mods are available for 1.21.1:
- Continuity (connected textures)
- Sodium (render optimization)
- Indium (fabric rendering API)
- Athena (connected textures)
- Axiom (world editor)

Update mixin configuration to remove references to disabled compat
mixins to prevent runtime crashes. These will be re-enabled in
future updates once the integration mods are updated.

sennecools committed:
docs: add comprehensive porting notes for 1.21.1 migration

Add detailed documentation covering all changes made during the port
from 1.20.4 to 1.21.1, including:
- Complete list of API changes and how they were addressed
- Build system requirements (Java 21)
- Testing status for all features
- Future work (Continuity integration)
- Build instructions

sennecools commented:
You probably want to make a new 1.21.1 branch instead of merging it into the main

DriHut commented:
Hello, very happy you took the time to do that. I will review this in the weekend and will indeed make a specific branch for that. If I have some more time I'll bring the fix for continuity I had made locally 😓.

sennecools committed:
feat: port Continuity compatibility to 3.0.0+1.21

Updates Continuity integration for the stable 3.0.0 release:
- Update class names from CTM to Ctm (case change)
- Update method signatures with new appearanceState and renderContext parameters
- Add null renderContext handling for DynamicBakedModel.computeQuads()
- Implement custom culling for transparent blocks (glass textures and frames)
- Re-enable RebakedModel support for CTM-processed models
- Support culling between framed blocks and vanilla blocks

Compatibility Status:
- Sodium: Compatible with 0.6.0+ (Indium no longer required)
- Continuity: Compatible with 3.0.0+1.21 (connected textures working)
- Indium: Not required on Sodium 0.6.0+ (native Fabric Rendering API support)

All connected textures functionality has been tested and verified working
with the latest stable versions of both Sodium and Continuity.

sennecools commented:
I was actually already working on continuity myself 😅 I think I got everything working, though I haven’t tested every edge case yet your implementation might even be better. I committed it anyway so you can take a look.
At the moment I don’t have plans to support the axiom compat, since I don’t use that  on my own server. My main goal was just to get the framed blocks working on 1.21.1 for my setup.

DriHut commented:
That's perfect. Makes less work for me 💪. I'm sure lots of people will be content with that. I get you didn't do the axiom compat since it was and still is a pain in the butt honestly 😓. As for Athena should be the same as Continuity in general, except the mod changed name I think. I'll see that in detail this weekend hopefully. So that people can finally play with it on 1.21.1 thanks to you.

sennecools commented:
I was looking into Athena for Chipped and Chisel, but it seems like Athena 4.0.2 uses Java 25, so I can’t compile against it in my Java 21 setup 😅
From what I can tell, Chipped and Chisel can still interact with Athena at runtime because Athena targets Java 21 bytecode, and the connected textures hook into runtime behavior rather than requiring direct compilation?
If you get a chance to look at it this weekend, I’d love any pointers on how you handled it or if there’s a way to add a runtime-only compat for the framed blocks. I’ll try a bit more on my own, but I’ve never done something like this before.

sennecools committed:
feat: extend compatibility to Minecraft 1.21.0 and 1.21.1

Changed version constraint from exact match to range to support both
1.21.0 and 1.21.1. The mod is fully compatible with both versions.

Version constraint updated:
- Before: "minecraft": "1.21.1" (exact match)
- After: "minecraft": ">=1.21 <=1.21.1" (range)

Why not 1.21.2+:
Starting with Minecraft 1.21.2, Mojang changed how block states are
initialized. Calling setDefaultState() in the block constructor now
throws "Block id not set" because the block ID is assigned during
registration, not before. This breaks all ReFramed blocks which set
their default LIGHT state in the constructor.

Additionally, 1.21.5+ has mixin field mapping changes where internal
fields like field_1760 (MinecraftClient) and field_22161 (WallBlock)
were renamed or relocated, causing InvalidAccessorException errors.

Known Issues:
- On 1.21.0, connected textures may show a small border on first load
  that shouldn't be there. This resolves itself after a relog.

DriHut committed:
added sennecools to authors
